xdsl/dialects/builtin.py

```diff
@@ -541,6 +541,71 @@ def verify(self, attr: Attribute) -> None:
                                | UnrankedTensorType[_VectorOrTensorElem])
 
 
+@dataclass(init=False)
+class VectorRankConstraint(AttrConstraint):
+    """
+    Constrain a vector to be of a given rank.
+    """
+
+    expected_rank: int
+    """The expected vector rank."""
+
+    def __init__(self, expected_rank: int):
+        self.expected_rank = expected_rank
+
+    def verify(self, vector: Attribute) -> None:
+        if not isinstance(vector, VectorType):
+            raise VerifyException(f"{vector} should be of type VectorType.")
+        if vector.get_num_dims() != self.expected_rank:
+            raise VerifyException(
+                f"Expected vector rank to be {self.expected_rank}, got {vector.get_num_dims()}."
+            )
+
+
+@dataclass(init=False)
+class VectorBaseTypeConstraint(AttrConstraint):
+    """
+    Constrain a vector to be of a given base type.
+    """
+
+    expected_type: ParameterDef[_VectorTypeElems]
+    """The expected vector base type."""
+
+    def __init__(self, expected_type: ParameterDef[_VectorTypeElems]):
+        self.expected_type = expected_type
+
+    def verify(self, vector: Attribute) -> None:
+        if not isinstance(vector, VectorType):
+            raise VerifyException(f"{vector} should be of type VectorType.")
+        if vector.element_type != self.expected_type:
+            raise VerifyException(
+                f"Expected vector type to be {self.expected_type}, got {vector.element_type}."
+            )
+
+
+@dataclass(init=False)
+class VectorBaseTypeAndRankConstraint(AttrConstraint):
+    """
+    Constrain a vector to be of a given rank and base type.
+    """
+
+    vector_base_type_verifier: VectorBaseTypeConstraint
```

Comment:
Again, I feel like it would be cool to find a way to build this in a more declarative way. So basically something like (AllOf)[https://github.com/xdslproject/xdsl/blob/main/xdsl/irdl.py#L130]. Maybe you can leverage that?

Reply:
Yes, this class is exactly an `AllOf([VectorBaseTypeConstraint(...), VectorRankConstraint(...)])`. I guess we could just make this class a function?

Reply:
My opinion is to express it as `AllOf([VectorBaseTypeConstraint(...), VectorRankConstraint(...)])` directly instead of developing syntactic sugar for this. To me, above version is more expressive and easily extensible than the other alternative. Let me know what you think of this.

Reply:
I think there is an argument to be had about how we want to do this exactly. In particular, there is "how users should do it" and "what we want to provide for the builtin dialects". IMO, this version might be the right thing for the builtin dialects, but it should not be what users define.